src/erasure.rs

```diff
@@ -161,263 +168,418 @@ pub fn decode_blocks(data: &mut [&mut [u8]], coding: &[&[u8]], erasures: &[i32])
+            if coding_wl.set_coding().is_err() {
+                return Err(ErasureError::EncodeError);
             }
 
-            generate_coding_blocks(coding_ptrs.as_mut_slice(), &data_ptrs)?;
-            debug!(
-                "consumed: {} data: {}:{} coding: {}:{}",
-                consumed,
-                block_start,
-                block_start + NUM_DATA,
-                coding_start,
-                coding_end
+            coding_blobs.push(coding.clone());
+        }
+
+        let mut data_locks = Vec::with_capacity(NUM_DATA);
+        for b in &data_blobs {
+            data_locks.push(
+                b.write()
+                    .expect("'data_locks' write lock in pub fn generate_coding"),
+            );
+        }
+
+        let mut data_ptrs: Vec<&[u8]> = Vec::with_capacity(NUM_DATA);
+        for (i, l) in data_locks.iter_mut().enumerate() {
+            trace!("{:x} i: {} data: {}", debug_id, i, l.data[0]);
+            data_ptrs.push(&l.data[..max_data_size]);
+        }
+
+        let mut coding_locks = Vec::with_capacity(NUM_CODING);
+        for b in &coding_blobs {
+            coding_locks.push(
+                b.write()
+                    .expect("'coding_locks' arr in pub fn generate_coding"),
             );
-            block_start += NUM_CODED;
         }
+
+        let mut coding_ptrs: Vec<&mut [u8]> = Vec::with_capacity(NUM_CODING);
+        for (i, l) in coding_locks.iter_mut().enumerate() {
+            trace!(
+                "{:x} i: {} coding: {} size: {}",
+                debug_id,
+                i,
+                l.data[0],
+                max_data_size
+            );
+            coding_ptrs.push(&mut l.data_mut()[..max_data_size]);
+        }
+
+        generate_coding_blocks(coding_ptrs.as_mut_slice(), &data_ptrs)?;
+        debug!(
+            "{:x} start_idx: {} data: {}:{} coding: {}:{}",
+            debug_id, start_idx, block_start, block_end, coding_start, block_end
+        );
+        block_start = block_end;
     }
     Ok(())
 }
 
-// Recover missing blocks into window
-//   missing blocks should be None, will use re
-//   to allocate new ones. Returns err if not enough
-//   coding blocks are present to restore
+// examine the window beginning at block_start for missing or
+//  stale (based on block_start_idx) blobs
+// if a blob is stale, remove it from the window slot
+//  side effect: block will be cleaned of old blobs
+fn find_missing(
+    debug_id: u64,
+    block_start_idx: u64,
+    block_start: usize,
+    window: &mut [WindowSlot],
+    recycler: &BlobRecycler,
+) -> (usize, usize) {
+    let mut data_missing = 0;
+    let mut coding_missing = 0;
+    let block_end = block_start + NUM_DATA;
+    let coding_start = block_start + NUM_DATA - NUM_CODING;
+
+    // count missing blobs in the block
+    for i in block_start..block_end {
+        let idx = (i - block_start) as u64 + block_start_idx;
+        let n = i % window.len();
+
+        fn is_missing(
```

Comment:
This is complex enough to be a top-level function, where it's easy to test each of those branches.